src/main/scala/com/databricks/spark/avro/DefaultSource.scala

```diff
@@ -168,7 +193,16 @@ private[avro] class DefaultSource extends FileFormat with DataSourceRegister {
           // Used to convert `Row`s containing data columns into `InternalRow`s.
           private val encoderForDataColumns = RowEncoder(requiredSchema)
 
-          override def hasNext: Boolean = reader.hasNext
+          private[this] var completed = false
+
+          override def hasNext: Boolean = {
+            val r = reader.hasNext
```

Comment:
Is it ok to call hasNext on a closed reader ? If not maybe you should check completed first. Also what happens if the reader is closed via the task completion callback above.